### PR TITLE
Use fixtures in pandas/tests/base

### PR DESCRIPTION
**Pull request by SaturnFromTitan**
part of #23877
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

**Note:** The diff is a bit inflated. Most changes are just indentation because for loops are replaced by a parametrized fixture.

### CODE REVIEW

File: pandas/conftest.py [hunk 1]
```diff
@@ -978,6 +978,7 @@ def __len__(self):
 
 @pytest.fixture(params=indices_dict.keys())
 def indices(request):
+    """ Fixture for all kinds of indices. """
```
Comment (jbrockmendel) [inline pandas/conftest.py]: Can this specify that the indices are "simple" not likely to cover corner cases, e.g. no names, no NaTs, no values near implementation bounds...

File: pandas/tests/base/test_ops.py [hunk 2]
```diff
@@ -137,227 +137,233 @@ def setup_method(self, method):
...
-            o = orig.copy()
-            klass = type(o)
-            values = o._values
-
-            if isinstance(values, Index):
-                # reset name not to affect latter process
-                values.name = None
+    def test_value_counts_unique_nunique(self, index_or_series_obj):
+        orig = index_or_series_obj
+        o = orig.copy()
```
Comment (WillAyd) [inline pandas/tests/base/test_ops.py]: don't technically need to copy with fixtures

File: pandas/tests/base/test_ops.py [hunk 3]
```diff
@@ -137,227 +137,233 @@ def setup_method(self, method):
...
-                o = klass(rep, index=idx, name="a")
-
-            # check values has the same dtype as the original
-            assert o.dtype == orig.dtype
-
-            expected_s = Series(
-                range(10, 0, -1), index=expected_index, dtype="int64", name="a"
-            )
+        if orig.duplicated().any():
+            pytest.skip("test implementation breaks with duplicated values")
```
Comment (WillAyd) [inline pandas/tests/base/test_ops.py]: Is this a bug?

File: pandas/tests/base/test_ops.py [hunk 4]
```diff
@@ -137,227 +137,233 @@ def setup_method(self, method):
...
-                    o = o.copy()
-                    o[0:2] = pd.NaT
-                    values = o._values
-
-            elif needs_i8_conversion(o):
-                values[0:2] = iNaT
-                values = o._shallow_copy(values)
+    def test_value_counts_unique_nunique_null(self, null_obj, index_or_series_obj):
+        orig = index_or_series_obj
+        o = orig.copy()
```
Comment (WillAyd) [inline pandas/tests/base/test_ops.py]: same comment on copy

**Review by WillAyd — COMMENTED** [review]
looks generally good (I think the naming of the fixture is a little strange, though admittedly don't have anything better to offer atm)

A few questions

File: pandas/tests/base/test_ops.py (hunk 3 above)
Comment (SaturnFromTitan) [reply pandas/tests/base/test_ops.py]: Not in the application code, but rather that the test isn't written properly. I wanted to focus on using the fixtures instead of rewriting the tests for now.
-> I'd leave fixing this to a follow-up PR if that's ok.

File: pandas/tests/base/test_ops.py (hunk 2 above)
Comment (SaturnFromTitan) [reply pandas/tests/base/test_ops.py]: You do, because `o` is being mutated, but we need to refer to `orig` later on in the test

File: pandas/tests/base/test_ops.py (hunk 4 above)
Comment (SaturnFromTitan) [reply pandas/tests/base/test_ops.py]: same as above

**Comment by SaturnFromTitan** [comment]
I think I addressed all your comments here @WillAyd

File: pandas/tests/indexes/common.py [hunk 5]
```diff
@@ -514,12 +514,12 @@ def test_union_base(self, indices):
 
     @pytest.mark.parametrize("sort", [None, False])
     def test_difference_base(self, sort, indices):
-        if isinstance(indices, CategoricalIndex):
-            return
-
         first = indices[2:]
         second = indices[:4]
-        answer = indices[4:]
+        if isinstance(indices, CategoricalIndex) or indices.is_boolean():
```
Comment (WillAyd) [inline pandas/tests/indexes/common.py]: Is this change related?

File: pandas/tests/base/test_ops.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/tests/base/test_ops.py]: @SaturnFromTitan you're very good about following up on things, but things slip through the cracks, so let's make sure this is an xfail with a helpful comment describing what needs to be fixed

File: pandas/conftest.py [hunk 6]
```diff
@@ -978,6 +978,15 @@ def __len__(self):
...
 def indices(request):
+    """
+    Fixture for many "simple" kinds of indices.
+
+    These indices are unlikely to cover corner cases, e.g.
+        - no names
+        - no NaTs/NaNs
+        - no values near implementation bounds
+        - ...
+    """
```
Comment (jbrockmendel) [inline pandas/conftest.py]: this is great, thanks

File: pandas/tests/indexes/common.py [hunk 7]
```diff
@@ -514,12 +514,12 @@ def test_union_base(self, indices):
 
     @pytest.mark.parametrize("sort", [None, False])
```
Comment (jbrockmendel) [inline pandas/tests/indexes/common.py]: not for this PR, but we have a fixture in tests.indexes.interval.test_setops for this

File: pandas/tests/indexes/common.py (hunk 5 above)
Comment (SaturnFromTitan) [reply pandas/tests/indexes/common.py]: Yes, it is. To make the indices fixture more useful I enhanced its length from 2 to 10 entries, which impacts the expected value here. In the same way, I could provide a proper "expected" for `CategoricalIndex` instead of just skipping it.

File: pandas/conftest.py [hunk 8]
```diff
@@ -995,6 +1004,14 @@ def _create_series(index):
 }
 
 
+@pytest.fixture(params=_series.keys())
+def series_with_differing_index(request):
```
Comment (jbrockmendel) [inline pandas/conftest.py]: not sure if this is more idiomatic (might actually be less clear), but i think this could be

```
@pytest.fixture
def series_with_differing_index(indices):
     return _create_series(indices)
```

File: pandas/tests/base/test_ops.py [hunk 9]
```diff
@@ -137,227 +137,233 @@ def setup_method(self, method):
         self.is_valid_objs = self.objs
         self.not_valid_objs = []
 
-    def test_none_comparison(self):
+    def test_none_comparison(self, series_with_differing_index):
```
Comment (jbrockmendel) [inline pandas/tests/base/test_ops.py]: series_with_differing_index makes me think "differing from what"?  i think Will had a similar issue

File: pandas/tests/base/test_ops.py [hunk 10]
```diff
@@ -137,227 +137,233 @@ def setup_method(self, method):
...
+        o = series_with_differing_index
+        if isinstance(o.index, IntervalIndex):
+            pytest.skip("IntervalIndex is immutable")
```
Comment (jbrockmendel) [inline pandas/tests/base/test_ops.py]: what does immutability have to do with this?  all indexes should be immutable

File: pandas/tests/base/test_ops.py [hunk 11]
```diff
@@ -137,227 +137,233 @@ def setup_method(self, method):
...
+        if len(o) < 1:
+            pytest.skip("Test doesn't make sense on empty data")
```
Comment (jbrockmendel) [inline pandas/tests/base/test_ops.py]: does this occur?  this seems like of like `hypothesis`'s `assume(len(o) >= 1)`

File: pandas/tests/base/test_ops.py (hunk 9 above)
Comment (SaturnFromTitan) [reply pandas/tests/base/test_ops.py]: What about `series_with_changing_index`?

File: pandas/tests/base/test_ops.py (hunk 11 above)
Comment (SaturnFromTitan) [reply pandas/tests/base/test_ops.py]: Yes it does, as we have an [empty in the indices fixture](https://github.com/pandas-dev/pandas/blob/master/pandas/conftest.py#L973)

File: pandas/tests/base/test_ops.py (hunk 10 above)
Comment (SaturnFromTitan) [reply pandas/tests/base/test_ops.py]: I'll have to check. I think the IntervalIndex doesn't implement `.loc` or something like that.

File: pandas/tests/base/test_ops.py [hunk 12]
```diff
@@ -137,227 +137,233 @@ def setup_method(self, method):
...
+        with pytest.raises(ValueError, match=msg):
+            o.item()  # len > 1
 
-            assert o.ndim == 1
-            assert o.size == len(o)
+        assert o.ndim == 1
+        assert o.size == len(o)
 
         assert Index([1]).item() == 1
         assert Series([1]).item() == 1
```
Comment (jbrockmendel) [inline pandas/tests/base/test_ops.py]: these last two lines look like they belong somewhere else

File: pandas/tests/base/test_ops.py (hunk 11 above)
Comment (jbrockmendel) [reply pandas/tests/base/test_ops.py]: thanks

File: pandas/tests/base/test_ops.py (hunk 10 above)
Comment (jbrockmendel) [reply pandas/tests/base/test_ops.py]: just a hunch, id imagine its like the datetime64 case where it raises for inequalities (in which case it can be added to the special case for datetime64 below)

File: pandas/tests/base/test_ops.py (hunk 9 above)
Comment (jbrockmendel) [reply pandas/tests/base/test_ops.py]: `f8series_any_simple_index`?

**Comment by jbrockmendel** [comment]
@SaturnFromTitan if its easy for you, can you point out sections of the diff that are effectively just-de-indentation, should make it easier to focus review

File: pandas/tests/base/test_ops.py (hunk 9 above)
Comment (SaturnFromTitan) [reply pandas/tests/base/test_ops.py]: What does the f8 stand for?

Comment (jbrockmendel) [reply pandas/tests/base/test_ops.py]: float64 dtype

Comment (SaturnFromTitan) [reply pandas/tests/base/test_ops.py]: Ok, I'll go with it then

File: pandas/tests/indexes/common.py (hunk 7 above)
Comment (SaturnFromTitan) [reply pandas/tests/indexes/common.py]: Created a follow-up issue: #32183

File: pandas/tests/base/test_ops.py (hunk 12 above)
Comment (SaturnFromTitan) [reply pandas/tests/base/test_ops.py]: Indeed. I wouldn't want to address it in this PR yet though

File: pandas/tests/base/test_ops.py (hunk 10 above)
Comment (SaturnFromTitan) [reply pandas/tests/base/test_ops.py]: No, it's failing on line 149 `o[0] = np.nan` already. The assignment fails with
```
AttributeError: 'pandas._libs.interval.IntervalTree' object has no attribute 'get_loc'
```

I added a comment and adjusted the skip message. Please let me know if you have a better idea

File: pandas/tests/base/test_ops.py [hunk 13]
```diff
@@ -137,227 +137,238 @@ def setup_method(self, method):
         self.is_valid_objs = self.objs
         self.not_valid_objs = []
 
-    def test_none_comparison(self):
+    def test_none_comparison(self, f8series_any_simple_index):
```
Comment (SaturnFromTitan) [inline pandas/tests/base/test_ops.py]: Fixed indentation and added skips

File: pandas/tests/base/test_ops.py [hunk 14]
```diff
@@ -137,227 +137,238 @@ def setup_method(self, method):
...
+            tm.assert_numpy_array_equal(
+                result.astype(object), orig._values.astype(object)
+            )
+        else:
+            tm.assert_numpy_array_equal(result, orig.values)
+            assert result.dtype == orig.dtype
 
-            assert o.nunique() == len(np.unique(o.values))
+        # dropna=True would break for MultiIndex
+        assert o.nunique(dropna=False) == len(np.unique(o.values))
```
Comment (SaturnFromTitan) [inline pandas/tests/base/test_ops.py]: added the dropna=False and a comment explaining why

File: pandas/tests/base/test_ops.py [hunk 15]
```diff
@@ -137,227 +137,238 @@ def setup_method(self, method):
...
 
-                    result = o < None
-                    assert not result.iat[0]
-                    assert not result.iat[1]
+            result = o < None
+            assert not result.iat[0]
+            assert not result.iat[1]
 
-    def test_ndarray_compat_properties(self):
+    def test_ndarray_compat_properties(self, index_or_series_obj):
```
Comment (SaturnFromTitan) [inline pandas/tests/base/test_ops.py]: Just fixed indentation

File: pandas/tests/base/test_ops.py [hunk 16]
```diff
@@ -137,227 +137,238 @@ def setup_method(self, method):
...
-                indices = np.repeat(np.arange(len(o)), range(1, len(o) + 1))
-                rep = values.take(indices)
-                o = klass(rep, index=idx, name="a")
-
-            # check values has the same dtype as the original
-            assert o.dtype == orig.dtype
-
-            expected_s = Series(
-                range(10, 0, -1), index=expected_index, dtype="int64", name="a"
+    def test_value_counts_unique_nunique(self, index_or_series_obj):
```
Comment (SaturnFromTitan) [inline pandas/tests/base/test_ops.py]: Fixed indentation and added xfail. All other changes are marked with comments

File: pandas/tests/base/test_ops.py [hunk 17]
```diff
@@ -137,227 +137,238 @@ def setup_method(self, method):
...
+            # take-based repeat
+            indices = np.repeat(np.arange(len(o)), range(1, len(o) + 1))
+            rep = values.take(indices)
+            o = klass(rep, index=idx)
+
+        # check values has the same dtype as the original
+        assert o.dtype == orig.dtype
+
+        expected_s = Series(
+            range(len(orig), 0, -1), index=expected_index, dtype="int64"
```
Comment (SaturnFromTitan) [inline pandas/tests/base/test_ops.py]: Made the length of range dynamic as not all values of `index_or_series_obj ` share the same length

File: pandas/tests/base/test_ops.py [hunk 18]
```diff
@@ -137,227 +137,238 @@ def setup_method(self, method):
...
-                    values = o._shallow_copy(v)
-                else:
-                    o = o.copy()
-                    o[0:2] = pd.NaT
-                    values = o._values
-
-            elif needs_i8_conversion(o):
-                values[0:2] = iNaT
-                values = o._shallow_copy(values)
+    def test_value_counts_unique_nunique_null(self, null_obj, index_or_series_obj):
```
Comment (SaturnFromTitan) [inline pandas/tests/base/test_ops.py]: Fixed indentation and added skips. All other changes are marked with comments

File: pandas/tests/base/test_ops.py [hunk 19]
```diff
@@ -137,227 +137,238 @@ def setup_method(self, method):
...
+        if isinstance(o, Index):
+            tm.assert_numpy_array_equal(pd.isna(o), nanloc)
+        else:
+            exp = Series(nanloc, o.index, name="a")
+            tm.assert_series_equal(pd.isna(o), exp)
+
+        expected_data = list(range(num_values, 2, -1))
+        expected_data_na = expected_data.copy()
+        if expected_data_na:
+            expected_data_na.append(3)
```
Comment (SaturnFromTitan) [inline pandas/tests/base/test_ops.py]: - length of expected_data is now dynamic
- using append instead of inline if/else statement

File: pandas/tests/base/test_ops.py [hunk 20]
```diff
@@ -137,227 +137,238 @@ def setup_method(self, method):
...
+            tm.assert_extension_array_equal(result[1:], values[2:])
+            assert result[0] is pd.NaT
+        elif len(o) > 0:
+            tm.assert_numpy_array_equal(result[1:], values[2:])
+
+            assert pd.isna(result[0])
+            assert result.dtype == orig.dtype
+
+        assert o.nunique() == max(0, num_values - 2)
+        assert o.nunique(dropna=False) == max(0, num_values - 1)
```
Comment (SaturnFromTitan) [inline pandas/tests/base/test_ops.py]: - comparing against dynamic value depending on the length of the fixture value
- using `max(0, ...)` so tests don't break with the empty index

**Comment by SaturnFromTitan** [comment]
I addressed all your comments @jbrockmendel 
Also, I added comments for all real changes in the tests to aid the review

File: pandas/conftest.py [hunk 21]
```diff
@@ -995,6 +1004,14 @@ def _create_series(index):
 }
 
 
+@pytest.fixture
+def f8series_any_simple_index(indices):
```
Comment (jreback) [inline pandas/conftest.py]: what is the purpose here? this name is impossible to grok

File: pandas/tests/base/test_ops.py [hunk 22]
```diff
@@ -137,227 +137,238 @@ def setup_method(self, method):
         self.is_valid_objs = self.objs
         self.not_valid_objs = []
 
-    def test_none_comparison(self):
+    def test_none_comparison(self, f8series_any_simple_index):
+        o = f8series_any_simple_index
```
Comment (jreback) [inline pandas/tests/base/test_ops.py]: call this something else, o is non-obvious (series would be better)

File: pandas/tests/base/test_ops.py [hunk 23]
```diff
@@ -137,227 +137,238 @@ def setup_method(self, method):
...
-                    result = o < None
-                    assert not result.iat[0]
-                    assert not result.iat[1]
+            result = o < None
+            assert not result.iat[0]
+            assert not result.iat[1]
 
-    def test_ndarray_compat_properties(self):
+    def test_ndarray_compat_properties(self, index_or_series_obj):
+        o = index_or_series_obj
```
Comment (jreback) [inline pandas/tests/base/test_ops.py]: o -> obj

File: pandas/tests/base/test_ops.py [hunk 24]
```diff
@@ -137,227 +137,238 @@ def setup_method(self, method):
...
-                o = klass(rep, index=idx, name="a")
-
-            # check values has the same dtype as the original
-            assert o.dtype == orig.dtype
-
-            expected_s = Series(
-                range(10, 0, -1), index=expected_index, dtype="int64", name="a"
+    def test_value_counts_unique_nunique(self, index_or_series_obj):
+        orig = index_or_series_obj
+        o = orig.copy()
```
Comment (jreback) [inline pandas/tests/base/test_ops.py]: o -> obj

File: pandas/tests/base/test_ops.py (hunk 18 above)
Comment (jreback) [inline pandas/tests/base/test_ops.py]: change to use nulls_fixtures

File: pandas/tests/base/test_ops.py [hunk 25]
```diff
@@ -137,227 +137,238 @@ def setup_method(self, method):
...
-                values[0:2] = iNaT
-                values = o._shallow_copy(values)
+    def test_value_counts_unique_nunique_null(self, null_obj, index_or_series_obj):
+        orig = index_or_series_obj
+        o = orig.copy()
+        klass = type(o)
+        values = o._ndarray_values
+        num_values = len(orig)
+
+        if not allow_na_ops(o):
```
Comment (jreback) [inline pandas/tests/base/test_ops.py]: @WillAyd don't we have a way to avoid using skips *inside* the test itself and rather do this externally? I don't recall exactly

File: pandas/tests/base/test_ops.py [hunk 26]
```diff
@@ -137,227 +137,238 @@ def setup_method(self, method):
...
             else:
-                tm.assert_numpy_array_equal(result[1:], values[2:])
-
-                assert pd.isna(result[0])
-                assert result.dtype == orig.dtype
+                expected_index = Index(values)
+            expected_index.name = None
+            o = o.repeat(range(1, len(o) + 1))
+            o.name = "a"
+
```
Comment (jreback) [inline pandas/tests/base/test_ops.py]: future PR, let's break these huge tests up

**Review by jreback — CHANGES_REQUESTED** [review]
looks good, some comments. ping on green.

File: pandas/conftest.py (hunk 21 above)
Comment (SaturnFromTitan) [reply pandas/conftest.py]: Any ideas for a better alternative @jreback?

@jbrockmendel and I already brainstormed a bit [in these comments](https://github.com/pandas-dev/pandas/pull/32046#discussion_r382704106), but couldn't come up anything something better...

Other ideas:
- `series_with_changing_index`
- `series_with_simple_index`
- ...

File: pandas/tests/base/test_ops.py (hunk 18 above)
Comment (SaturnFromTitan) [reply pandas/tests/base/test_ops.py]: I'd take care of this in a follow-up PR. The tests are breaking with `NaT`, so it would require some extra effort to fix/skip it, but the PR is already pretty big

File: pandas/conftest.py (hunk 21 above)
Comment (jreback) [reply pandas/conftest.py]: series_with_simple_index is more indicative

File: pandas/tests/base/test_ops.py (hunk 25 above)
Comment (SaturnFromTitan) [reply pandas/tests/base/test_ops.py]: I think you can only do this via the `pytest.mark.skip` decorator. But this approach doesn't work here, because we need a condition on the parametrized fixture value. A quick Google search also didn't bring up anything better

**Comment by SaturnFromTitan** [comment]
@jreback CI is green and I addressed all your comments

File: pandas/tests/base/test_ops.py [hunk 27]
```diff
@@ -137,227 +137,238 @@ def setup_method(self, method):
...
+        if not allow_na_ops(obj):
+            pytest.skip("type doesn't allow for NA operations")
+        elif isinstance(orig, (pd.CategoricalIndex, pd.IntervalIndex)):
+            pytest.skip(f"values of {klass} cannot be changed")
+        elif isinstance(orig, pd.MultiIndex):
+            pytest.skip("MultiIndex doesn't support isna")
+
+        # special assign to the numpy array
+        if is_datetime64tz_dtype(obj):
+            if isinstance(obj, DatetimeIndex):
```
Comment (jreback) [inline pandas/tests/base/test_ops.py]: this can be simplified; the else path *should* work for all

File: pandas/tests/base/test_ops.py [hunk 28]
```diff
@@ -137,227 +137,238 @@ def setup_method(self, method):
...
+                obj = obj.copy()
+                obj[0:2] = pd.NaT
+                values = obj._values
 
-            # create repeated values, 'n'th element is repeated by n+1
-            # times
-            if isinstance(o, (DatetimeIndex, PeriodIndex)):
-                expected_index = o.copy()
-                expected_index.name = None
+        elif needs_i8_conversion(obj):
```
Comment (jreback) [inline pandas/tests/base/test_ops.py]: this is also duplicative here

File: pandas/tests/base/test_ops.py [hunk 29]
```diff
@@ -137,227 +137,238 @@ def setup_method(self, method):
...
-                tm.assert_index_equal(result, Index(values[1:], name="a"))
-            elif is_datetime64tz_dtype(o):
-                # unable to compare NaT / nan
-                tm.assert_extension_array_equal(result[1:], values[2:])
-                assert result[0] is pd.NaT
+            # attach name to klass
+            obj = klass(values.repeat(range(1, len(obj) + 1)))
+            obj.name = "a"
+        else:
+            if isinstance(obj, DatetimeIndex):
```
Comment (jreback) [inline pandas/tests/base/test_ops.py]: this can't possibly be true (as the first if already catches this)

**Review by jreback — APPROVED** [review]
going to merge, but there are a number of weird constructs / duplications here, pls address these as a followup (or open an issue)

**Comment by jreback** [comment]
thanks. ideally very targeted PRs are best as we can merge quickly.

**Comment by SaturnFromTitan** [comment]
That's the plan @jreback I wanted to focus on fixturizing the tests before refactoring.

Seeing how this PR went, it apparently makes more sense to refactor right away and go test by test instead ✌️

File: pandas/tests/base/test_ops.py (hunk 27 above)
Comment (SaturnFromTitan) [reply pandas/tests/base/test_ops.py]: xref: #32205
